_posts/2022-08-04-calldata-tuple-reencoding-head-overflow-bug.md

```diff
@@ -0,0 +1,219 @@
+|    bool    |    uint     | bytes32[2] ||                   T                    |
+|            |             |            ||--------------------------||------------|
+|            |             |            ||          HEAD of T       || TAIL of T  |
+|            |             |            ||-------------+------------||------------|
+|            |             |            || offset of x | value of y || value of x |
+|            |             |            ||    uint     |   uint[3]  ||    bytes   |
+|------------+-------------+------------||-------------+------------||------------|
+|     1      |      2      |      6     ||      3      |     5      ||      4     |
+|------------+-------------+------------||-------------+------------||------------|
+```
+
+The encoding order was also affected.
+This is because for each component, the encoder writes both the head and the tail before moving on to the next one.
+This results in `c` being written after `b.x`, even though the former precedes the latter in the encoding.
+Similar situation happened inside `T` - when `y` is being written the data belonging to tail of `x` is already in place.
+
+## Effects of the Bug
+
+The aggressive cleanup combined with the non-linear encoding order of dynamically-sized types resulted in a bug where
+the initial 32 bytes of the tail could be overwritten during the cleanup of the last head component.
+
+The dynamically-sized tuple example above is affected and would result in the following incorrect output from `D.f()`:
+
+```
+0x0000000000000000000000000000000000000000000000000000000000000001 a
+0x0000000000000000000000000000000000000000000000000000000000000080 offset of b
+0x6100000000000000000000000000000000000000000000000000000000000000 c[0]
+0x6200000000000000000000000000000000000000000000000000000000000000 c[1]
+0x0000000000000000000000000000000000000000000000000000000000000000 offset of b.x       // Wrong value
+0x000000000000000000000000000000000000000000000000000000000000000b b.y[0]
+0x000000000000000000000000000000000000000000000000000000000000000c b.y[1]
+0x000000000000000000000000000000000000000000000000000000000000000d b.y[2]
+0x0000000000000000000000000000000000000000000000000000000000000000 b.x (length field)  // Wrong value
+0x6162636400000000000000000000000000000000000000000000000000000000 b.x (data)
+```
+
+Here, the cleanup performed for `b.y` would overwrite the length field of `b.x`.
+Similarly, the cleanup of `c` would overwrite the offset of `b.x`.
+
+The case with a statically-sized tuple was not affected because, in absence of the tail, the encoding order is linear.
+The encoder could still write past the end of the whole area reserved for the encoding but the compiler would be able
+to ensure that the memory past it had not been allocated for another purpose, making the cleanup perfectly safe.
+
+The bug would only affect encoding of values stored in calldata because that was the only situation where
```

Comment:
```suggestion
The bug only affects the encoding of values stored in calldata because that was the only situation where
```

Although as mentioned above, IIUC, only the statically-sized array causing the cleanup has to come from calldata...

Reply:
Done, though it was meant to be past tense, so I used `affected` (doesn't `would affect` really convey the same meaning in this case?)

As for the `calldata` part, still working on fixing that in all parts of the post.